Rpi 4.9.y adxl372 fifo

stefpopa opened the pull request:
This patch adds support for the adxl372 FIFO. In order to accomplish this, triggered buffers were used.
Therefore, adxl372 is configured to trigger the watermark interrupt when 170 sets of concurrent 3-axis data is available in the FIFO. This data along with the timestamp is pushed to the IIO device's buffer. Also, the FIFO is configured in peak detect and stream mode.

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -248,21 +275,44 @@ static const struct iio_chan_spec adxl372_channels[] = {
 	union {
 		__le16 regval;
-		u8 d8[512];
+		u8 d8[1024];
```
this change requires a bit more explanation and maybe it's own commit;
you're essentially doubling the data here

stefpopa replied:
When data is popped from the FIFO, it is stored in this buffer. The max FIFO size is 512 samples of 2 bytes each -> 1024 bytes. 
Before the FIFO was used, this buffer could have been much smaller. However, it is true that 512 bytes was too much.

commodo replied:
thanks for the info :)
i guess a comment in the commit would be fine ; unless there are other objections ;

stefpopa replied:
ack; will add this info in the commit message

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -248,21 +275,44 @@ static const struct iio_chan_spec adxl372_channels[] = {
+};
+
 struct adxl372_state {
-	struct spi_device		*spi;
```
hmm, if keeping things aligned here is a concern/desire, maybe you could compress some type names;
so `struct adxl372_interrupt_info` ==> `struct adxl372_int_info` and stuff like that

i would not re-indent stuff for 1-2 new fields;

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -422,6 +493,151 @@ static int adxl372_set_activity_threshold(struct adxl372_state *st,
+}
+
+static int adxl372_configure_fifo(struct adxl372_state *st,
+				  struct adxl372_fifo_config fifo_config)
```
convert to pointer ; otherwise this gets copied onto the stack

stefpopa commented on drivers/iio/accel/adxl372.c:
```diff
@@ -287,6 +337,27 @@ static int adxl372_spi_write_mask(struct adxl372_state *st,
+	 * data is not overwritten and stored out of order, at least one sample
+	 * set must be left in the FIFO after every read.
+	 */
+	*fifo_entries -= st->fifo_set_size;
```
if it is not modified inside adxl372_read_fifo() function, then it has to be changed again before it is pushed to buffers

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -422,6 +493,151 @@ static int adxl372_set_activity_threshold(struct adxl372_state *st,
+
+	ret = adxl372_get_status(st, &status1, &status2, &fifo_entries);
+	if (ret < 0)
+		return ret;
```
goto err

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -587,18 +815,61 @@ static int adxl372_probe(struct spi_device *spi)
 	return 0;
+
+fail_dev_register:
+fail_trigger_alloc:
```
maybe just use one label here;
or, are you planning to add more stuff in-between the labels later ?

stefpopa replied:
you are right, one label should be fine

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -248,21 +275,44 @@ static const struct iio_chan_spec adxl372_channels[] = {
+	enum adxl372_fifo_mode mode;
+	enum adxl372_fifo_format format;
+	u16 samples;
+} adxl372_fifo_default_config = {
```
Does a user have to re-compile the kernel to change this?

I think these should be configurable using the standard IIO API. The format depends on the selected channels. Samples can be configured using the buffer watermark.

Please have a look at the IIO HW FIFO ABI: https://github.com/analogdevicesinc/linux/blob/master/Documentation/ABI/testing/sysfs-bus-iio#L1378-L1470

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -422,6 +493,151 @@ static int adxl372_set_activity_threshold(struct adxl372_state *st,
+			if (ret < 0)
+				goto err;
+
+			for (i = 0; i < fifo_entries * 2;
```
Indentation is a bit to high here. Either restructure or use a helper function.

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,6 +761,10 @@ static int adxl372_raw(struct iio_dev *indio_dev,
 }
 
+static const struct iio_trigger_ops adxl372_trigger_ops = {
+	.owner = THIS_MODULE,
```
Is there nothing to do when triggered capture starts/stops? Like enable interrupts, etc.

lclausen-adi commented on drivers/iio/accel/adxl372.c:
```diff
@@ -587,18 +815,61 @@ static int adxl372_probe(struct spi_device *spi)
+	st->dready_trig->ops = &adxl372_trigger_ops;
+	st->dready_trig->dev.parent = &st->spi->dev;
+	iio_trigger_set_drvdata(st->dready_trig, indio_dev);
+	iio_trigger_register(st->dready_trig);
```
check return code

stefpopa commented:
changelog v1 -> v2:
* the number of samples is now configurable using the buffer watermark
* the format depends on the number of selected channels
* added const attributes `hwfifo_watermark_min` and `hwfifo_watermark_max`
* added device attribute `hwfifo_enabled`
* `adxl372_buffer_postenable()` configures the FIFO with the number of samples and the fifo format and sets the FIFO_FULL interrupt
* `adxl372_buffer_predisable() `disables interrupts and sets the FIFO mode to bypassed
* includes a `set_trigger_state` member in `adxl372_trigger_ops`  which sets or resets the interrupts by checking the state
* the return code of `iio_trigger_register` is checked
* `fail_dev_register` label has been removed as `fail_trigger_alloc` is enough
* `adxl372_trigger_handler` has been restructured 
* removed extra white spaces from` struct adxl372_state`
* the commit message explains why d8 array size was changed from 512 to 1024
* fifo_entries is not passed to of `adxl372_read_fifo` function as a pointer, instead its value is decremented inside the adxl372_trigger_handler before calling `adxl372_read_fifo`
* commit [9f4667776](https://github.com/torvalds/linux/commit/9f4667776c138df33c4107fcd8811aa9cb6cdcbe#diff-c54e394f36af612cad9870b561cac278) had to be backported in this PR as this change required the 
`iio_buffer_set_attrs`

commodo commented on drivers/iio/accel/bmc150-accel-core.c:
```diff
@@ -1632,7 +1632,8 @@ int bmc150_accel_core_probe(struct device *dev, struct regmap *regmap, int irq,
 		if (block_supported) {
 			indio_dev->modes |= INDIO_BUFFER_SOFTWARE;
 			indio_dev->info = &bmc150_accel_info_fifo;
-			indio_dev->buffer->attrs = bmc150_accel_fifo_attributes;
```
I'd reference the original upstream commit for this.

stefpopa replied:
done; edited the changelog

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,10 +710,164 @@ static int adxl372_raw(struct iio_dev *indio_dev,
+
+	enabled = st->fifo_mode;
+
+	return sprintf(buf, "%d\n", enabled);
```
i'm a bit unclear about the practice of the newline char here;
should it be added or not ?
@larsclausen

commodo replied:
[curios] is there a way to use `snprintf()` here ?

lclausen-adi replied:
newline is correct here.

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,10 +710,164 @@ static int adxl372_raw(struct iio_dev *indio_dev,
+	 * 256 sample sets of concurrent 2-axis data (user selectable)
+	 * 512 sample sets of single-axis data
+	 */
+	if ((st->watermark * fifo_set_size) > 512)
```
i'd use `ADXL372_FIFO_SIZE` here, since it's defined ;

stefpopa replied:
agree

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,10 +710,164 @@ static int adxl372_raw(struct iio_dev *indio_dev,
+	int ret;
+
+	if (state)
+		ret = adxl372_set_interrupts(st, ADXL372_INT1_MAP_FIFO_FULL_MSK, 0);
```
[nitpick]  this doesn't need to be changed ;
this could be compressed a bit:
```
u8 mask = state ? ADXL372_INT1_MAP_FIFO_FULL_MSK : 0;
return adxl372_set_interrupts(st, mask, 0);
```

lclausen-adi replied:
upstream people don't like the ternary operator

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -579,26 +906,80 @@ static int adxl372_probe(struct spi_device *spi)
 
-	ret = iio_device_register(indio_dev);
-	if (ret)
+	ret = iio_triggered_buffer_setup(indio_dev, NULL,
```
you could  `devm_iio_triggered_buffer_setup()` here ; 
then you can remove: `iio_triggered_buffer_cleanup(indio_dev);` as this resource becomes handled automatically at remove()

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -579,26 +906,80 @@ static int adxl372_probe(struct spi_device *spi)
 	iio_device_unregister(indio_dev);
+	if (st->dready_trig) {
+		iio_triggered_buffer_cleanup(indio_dev);
+		iio_trigger_unregister(st->dready_trig);
```
I think this can be removed here ;
`devm_iio_trigger_alloc()` would make this resource managed by the device;

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -579,26 +906,80 @@ static int adxl372_probe(struct spi_device *spi)
+	if (st->dready_trig == NULL) {
+		ret = -ENOMEM;
+		dev_err(&indio_dev->dev, "Failed to allocate iio trigger\n");
+		goto fail_trigger_alloc;
```
i think just doing `return -ENOMEM` here should be fine ;
this resource is managed by the device; and so are the ones above; so, return an err code should let the device handle them at remove() [or when to free them]

stefpopa commented:
changelog v3 -> v4:
* referenced the original upstream commit in the backported commit's description

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,10 +710,164 @@ static int adxl372_raw(struct iio_dev *indio_dev,
+	struct adxl372_state *st = iio_priv(indio_dev);
+	bool enabled;
+
+	enabled = st->fifo_mode;
```
if you want to downcast the type of `fifo_mode` to bool, i suggest doing  `enabled = !!st->fifo_mode;`

commodo replied:
tho, the use of the extra variable may not be needed

commodo commented on drivers/iio/accel/adxl372.c:
```diff
@@ -537,10 +710,164 @@ static int adxl372_raw(struct iio_dev *indio_dev,
+
+	watermark = st->watermark;
+
+	return sprintf(buf, "%d\n", watermark);
```
nitpick: you could just `return sprintf(buf, "%d\n", st->watermark);`
but it's fine to leave it as is

commodo replied:
ah....
i noticed that `st->watermark` is `u16`
so the assignment/cast makes sense to avoid a warning with `sprintf()`
so you'd need to use `return sprintf(buf, "%u\n", st->watermark);` here

but let's leave it as is

commodo commented:
otherwise looks good from my side ; 

we may want to backport the Travis CI stuff from master [later] once it gets merged;
otherwise we'll be spammed with build errs;

stefpopa commented:
changelog v4 -> v5:
* removed the unnecessary extra variables from `adxl372_get_fifo_enabled()` and `adxl372_get_fifo_watermark()`

commodo commented:
looks good from my side;
AFAICT, comments have been addressed ;
hopefully I did not miss any ;

if no objections, i will merge tomorrow afternoon
